Implement Event::WindowMoved for Win32

learn-more opened the pull request:
-   [x] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
-   [x] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
-   [x] Have you provided some example/test code for your changes?
-   [ ] If you have additional steps which need to be performed, please list them as tasks!


## Description

Add an event that will be raised after a window is moved (just like the size event works).
This has been discussed / promised multiple times before, at least:
https://en.sfml-dev.org/forums/index.php?topic=2914.msg19032
https://en.sfml-dev.org/forums/index.php?topic=3598.msg23525#msg23525
https://en.sfml-dev.org/forums/index.php?topic=19658.msg141513#msg141513


## Tasks

-   [ ] Tested on Linux
-   [x] Tested on Windows
-   [ ] Tested on macOS
-   [ ] Tested on iOS
-   [ ] Tested on Android

## How to test this PR?

<!-- Describe how to best test these changes. -->

<!-- Please provide a [minimal, complete and verifiable example](https://stackoverflow.com/help/mcve) if possible, you can use the following template as a start: -->

```cpp
#include <SFML/Graphics.hpp>

int main()
{
    sf::RenderWindow window(sf::VideoMode({1280, 720}), "Minimal, complete and verifiable example");

    while (window.isOpen())
    {
        while (const std::optional event = window.pollEvent())
        {
            if (event->is<sf::Event::Closed>())
                window.close();
            else if (const auto* moved = event->getIf<sf::Event::WindowMoved>())
                printf("New position: %d, %d\n", moved->position.x, moved->position.y);
        }

        window.clear();
        window.display();
    }
}
```

ChrisThrasher commented:
What is your use case for window moved events?

learn-more commented:
> What is your use case for window moved events?

I wanted to use this as a basis for writing desktop tools,
in which I sometimes use this as 'overlay' over the actual desktop:
![image](https://github.com/user-attachments/assets/33b33884-ee8f-467f-b094-fc52ccde0ef7)

If this is denied because there is no use for it that would be fair as well,
but considering it was said to 'probably come in sfml2' I assumed it was still something that had at least some interest.

binary1248 commented:
While I don't have anything against such functionality (other libraries provide similar), I still want to see at least one other operating system implementation. Designing the API around what a single operating system supports goes against the idea that SFML is a cross-platform library that provides portable abstractions.

It has to also be considered whether the event has to contain the new window position or not. While there might be valid use cases for being notified when the window has been moved, not everybody is going to be interested in its new position, and for those who are, they can just call `getPosition()` to get the value themselves.

learn-more commented:
> While I don't have anything against such functionality (other libraries provide similar), I still want to see at least one other operating system implementation. Designing the API around what a single operating system supports goes against the idea that SFML is a cross-platform library that provides portable abstractions.
> 
> It has to also be considered whether the event has to contain the new window position or not. While there might be valid use cases for being notified when the window has been moved, not everybody is going to be interested in its new position, and for those who are, they can just call `getPosition()` to get the value themselves.

The `WindowMoved` contains the position, because the `Resized` event contains the window size.
If you want me to change this, please let me know.

As for implementing others, I'll have to spin up a VM to get anything other than Windows running, this would probably be some linux distro.

binary1248 commented:
Normally I tend to only add data to a structure if I know the data will be often used, but in this case I can see the consistency argument when comparing to `Resized`. The overhead shouldn't be an issue since `Moved` events should happen relatively rarely during the lifetime of an application anyway. If anybody is having performance issues due to `Moved` events being too big I think they have other problems to solve first.

learn-more commented:
> Normally I tend to only add data to a structure if I know the data will be used often, but in this case I can see the consistency argument when comparing to `Resized`. The overhead shouldn't be an issue since `Moved` events should happen relatively rarely during the lifetime of an application anyway. If anybody is having performance issues due to `Moved` events being too big I think they have other problems to solve first.

Rebased on the latest master to fix the merge conflicts in `Event.test.cpp`
Added an X11 implementation, and I temporarily monkeypatched the `window` example to print the coordinates:

![image](https://github.com/user-attachments/assets/d27edf12-c38e-483f-b79f-94add0e832ab)

eXpl0it3r commented on include/SFML/Window/Event.hpp:
```diff
@@ -366,6 +375,7 @@ class Event
     ////////////////////////////////////////////////////////////
     std::variant<Closed,
                  Resized,
+                 WindowMoved,
```
Wonder if it should just be `Moved` given that it also isn't `WindowResized`, then again Resized doesn't stand in name conflict with say `MouseMoved`.

Any opinions on this? 🤔

binary1248 replied:
If I had to guess, `Resized` isn't `WindowResized` because it didn't require the disambiguation, whereas `Moved` does.

learn-more replied:
I chose this name because it is more explicit about what is being moved.
Imo resized should be renamed to WindowResized, but that would probably also require a typedef for backwards compatibility

ChrisThrasher commented on test/Window/Event.test.cpp:
```diff
@@ -50,14 +55,22 @@ TEST_CASE("[Window] sf::Event")
             const auto& resized = *event.getIf<sf::Event::Resized>();
             CHECK(resized.size == sf::Vector2u(1, 2));
         }
+        SECTION("Template constructor window move")
+        {
+            const sf::Event event = sf::Event::WindowMoved{{1, 2}};
+            CHECK(event.is<sf::Event::WindowMoved>());
+            CHECK(event.getIf<sf::Event::WindowMoved>());
+            const auto& windowMoved = *event.getIf<sf::Event::WindowMoved>();
+            CHECK(windowMoved.position == sf::Vector2i(1, 2));
+        }
```
```suggestion
```

kimci86 replied:
Chris is suggesting to delete those lines.

ChrisThrasher commented on test/Window/Event.test.cpp:
```diff
@@ -50,14 +55,22 @@ TEST_CASE("[Window] sf::Event")
     SECTION("Construction")
     {
-        SECTION("Template constructor")
+        SECTION("Template constructor resize")
```
```suggestion
        SECTION("Template constructor")
```

ChrisThrasher commented on test/Window/Event.test.cpp:
```diff
@@ -307,6 +329,12 @@ TEST_CASE("[Window] sf::Event")
         REQUIRE(mouseMoved);
         mouseMoved->position = sf::Vector2i(6, 9);
         CHECK(mouseMoved->position == sf::Vector2i(6, 9));
+
+        event               = sf::Event::WindowMoved{{1, 2}};
+        auto* windowMoved   = event.getIf<sf::Event::WindowMoved>();
+        REQUIRE(windowMoved);
+        windowMoved->position = sf::Vector2i(3, 4);
+        CHECK(windowMoved->position == sf::Vector2i(3, 4));
```
```suggestion
```

ChrisThrasher commented on src/SFML/Window/Unix/WindowImplX11.cpp:
```diff
@@ -1757,6 +1757,13 @@ bool WindowImplX11::processEvent(XEvent& windowEvent)
                 m_previousSize.x = windowEvent.xconfigure.width;
                 m_previousSize.y = windowEvent.xconfigure.height;
             }
+            if (windowEvent.xconfigure.x != m_previousPosition.x || windowEvent.xconfigure.y != m_previousPosition.y)
+            {
+                pushEvent(Event::WindowMoved{Vector2i(windowEvent.xconfigure.x, windowEvent.xconfigure.y)});
+
+                m_previousPosition.x = windowEvent.xconfigure.x;
+                m_previousPosition.y = windowEvent.xconfigure.y;
+            }
```
```suggestion
            const sf::Vector2i previousPosition(windowEvent.xconfigure.x, windowEvent.xconfigure.y);
            if (m_previousPosition != previousPosition)
            {
                pushEvent(Event::WindowMoved{previousPosition});
                m_previousPosition = previousPosition;
            }
```
